blog/content/post/2021-04-18-optiplex_part2.md

```diff
@@ -0,0 +1,249 @@
+is possible that the Environmental Controller keeps its configuration as long
+as standby power is supplied. So taking it into consideration and all my
+previous conclusions I have decided to enable the USB dongle debug instead of
+the onboard serial port. Why? Look at this:
+
+```C
+/* Should support 8250, 16450, 16550, 16550A type UARTs */
+
+/* Expected character delay at 1200bps is 9ms for a working UART
+ * and no flow-control. Assume UART as stuck if shift register
+ * or FIFO takes more than 50ms per character to appear empty.
+ *
+ * Estimated that inb() from UART takes 1 microsecond.
+ */
+#define SINGLE_CHAR_TIMEOUT	(50 * 1000)
+#define FIFO_TIMEOUT		(16 * SINGLE_CHAR_TIMEOUT)
+
+static int uart8250_can_tx_byte(unsigned int base_port)
+{
+	return inb(base_port + UART8250_LSR) & UART8250_LSR_THRE;
+}
+
+static void uart8250_tx_byte(unsigned int base_port, unsigned char data)
+{
+	unsigned long int i = SINGLE_CHAR_TIMEOUT;
+	while (i-- && !uart8250_can_tx_byte(base_port));
+	outb(data, base_port + UART8250_TBR);
+}
+```
+
+So the UART driver in coreboot checks whether a character can be transmitted
+over serial port and then transmits it. However, if the serial port is not
+functioning, the `while` loop may be reaching the defined timeout, which can be
+even up to 50ms. Now image: full debug coreboot output has over 50000
+characters. So a quick calculation: 50000 * 50ms = 2500000 ms = 2500s ~ 42min.
+So the whole boot process would take over 40 minutes, no wondering that I have
+assumed that the board is not starting at all. This is the first time I have
+encountered such issue with Super I/O where the serial port is not working
+despite having a known good driver. Just in case I stopped producing builds
+with onboard serial port as the main debug method. I bought FT232H which is
+supported by coreboot and used it as EHCI USB debug dongle. I was able to get
+the debug output from coreboot without worrying about AC loss or whatsoever.
+
+## Summary
+
+So the issue with board not booting after AC loss has been resolved, or at
+least located. It took me enormous amount of time to figure it out. But don't
+feel sorry, it's a daily firmware developer job to debug for many hours just to
+find a simple fix. And I kind of like it. It is a professional deviation I
+could say.
+
+In the next post I will disclose more details how I implemented the
+Environmental Controller driver for the Dell OptiPlex 9010. It should show some
+progress on making the fan control work and how I have achieved the current
+implementation level which is available in upstream coreboot repository. Also
+you will see how far one has to go to achieve a goal in firmware development.
+Stay tuned.
+
+If you think we can help in improving the security of your firmware or you
+looking for someone who can boost your product by leveraging advanced features
+of used hardware platform, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
```

Comment:
```suggestion
If you have a problem with the Environmental Controller (EC) of your platform, looking to
write EC debugging or updating utilities, we are one to discuss. Also, since [EC firmware ecosystem shifts to Zephyr](https://cfp.osfc.io/osfc2020/talk/SXYUQD/) 
and we are huge fans of Zephyr. We would be glad to implement Firmware or even Open Source Firmware for your EC. If you are interested, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
```